File: hw/saraguro/u3/HW27DesignPatternsPrt3/Strategy/src/main/java/ec/edu/espe/strategy/controller/SortingContext.java
package ec.edu.espe.strategy.controller;

import ec.edu.espe.strategy.model.BubbleSort;
import ec.edu.espe.strategy.model.InsertionSort;
import ec.edu.espe.strategy.model.QuickSort;
import ec.edu.espe.strategy.model.SortingStrategy;

/**
 *
 * @author Leidy Saraguro
 */

public class SortingContext {

    private SortingStrategy ss;

    public int[] sort(int data[]) {
        int size = data.length;
        ss = setSortStrategy(size);
        return ss.sort(data);
    }

    public SortingStrategy setSortStrategy(int n) {
        if (n > 0 && n < 30) {
            ss = new BubbleSort();
        } else if (n >= 30 && n < 100) {
            ss = new InsertionSort();
        } else {
            ss = new QuickSort();
        }
        return ss;
    }
}

File: exams/u3/mera/Exam3Mera/Exam3Mera/MeraExam/controller/sorting_context.py
from model.bubble_sort import BubbleSort
from model.insertion_sort import InsertionSort
from model.quick_sort import QuickSort
from model.mongodb import Database

class SortingContext:
    def __init__(self):
        self.strategy = None
        self.database = Database()

    def set_sort_strategy(self, data):
        size = len(data)
        if size < 2:
            raise ValueError("Array must have more than 1 element")
        elif 2 <= size <= 5:
            self.strategy = BubbleSort()
        elif 6 <= size <= 10:
            self.strategy = InsertionSort()
        else:
            self.strategy = QuickSort()

    def sort(self, data):
        self.set_sort_strategy(data)
        sorted_data = self.strategy.sort(data.copy())
        record = {
            "unsorted": ",".join(map(str, data)),
            "size": len(data),
            "sort algorithm": self.strategy.__class__.__name__,
            "sorted": ",".join(map(str, sorted_data))
        }
        self.database.save_record(record)
        return sorted_data, self.strategy.__class__.__name__
